cpp/src/filesystem/s3/multi_part_upload_s3_fs.cpp

```diff
@@ -0,0 +1,2293 @@
+                     const arrow::io::IOContext& io_context,
+                     const S3Path& path,
+                     const S3Options& options,
+                     const std::shared_ptr<const arrow::KeyValueMetadata>& metadata,
+                     const int64_t part_size)
+      : holder_(std::move(holder)),
+        io_context_(io_context),
+        path_(path),
+        metadata_(metadata),
+        default_metadata_(options.default_metadata),
+        background_writes_(options.background_writes),
+        part_upload_size_(part_size),
+        multi_part_upload_threshold_size_(part_size - 1) {}
+
+  ~CustomOutputStream() override {
+    // For compliance with the rest of the IO stack, Close rather than Abort,
+    // even though it may be more expensive.
+    CloseFromDestructor(this);
+  }
+
+  template <typename ObjectRequest>
+  arrow::Status SetMetadataInRequest(ObjectRequest* request) {
+    std::shared_ptr<const arrow::KeyValueMetadata> metadata;
+
+    if (metadata_ && metadata_->size() != 0) {
+      metadata = metadata_;
+    } else if (default_metadata_ && default_metadata_->size() != 0) {
+      metadata = default_metadata_;
+    }
+
+    bool is_content_type_set{false};
+    if (metadata) {
+      RETURN_NOT_OK(SetObjectMetadata(metadata, request));
+
+      is_content_type_set = metadata->Contains("Content-Type");
+    }
+
+    if (!is_content_type_set) {
+      // If we do not set anything then the SDK will default to application/xml
+      // which confuses some tools (https://github.com/apache/arrow/issues/11934)
+      // So we instead default to application/octet-stream which is less misleading
+      request->SetContentType("application/octet-stream");
+    }
+
+    return arrow::Status::OK();
+  }
+
+  std::shared_ptr<CustomOutputStream> Self() {
+    return std::dynamic_pointer_cast<CustomOutputStream>(shared_from_this());
+  }
+
+  arrow::Status CreateMultipartUpload() {
+    DCHECK(ShouldBeMultipartUpload());
+
+    ARROW_ASSIGN_OR_RAISE(auto client_lock, holder_->Lock());
+
+    // Initiate the multi-part upload
+    Aws::S3::Model::CreateMultipartUploadRequest req;
+    req.SetBucket(ToAwsString(path_.bucket));
+    req.SetKey(ToAwsString(path_.key));
+    RETURN_NOT_OK(SetMetadataInRequest(&req));
+
+    auto outcome = client_lock.Move()->CreateMultipartUpload(req);
+    if (!outcome.IsSuccess()) {
+      return arrow::Status::Invalid("When initiating multiple part upload for key '", path_.key, "' in bucket '",
+                                    path_.bucket + "': ", "CreateMultipartUpload", outcome.GetError());
+    }
+    multipart_upload_id_ = outcome.GetResult().GetUploadId();
+
+    return arrow::Status::OK();
+  }
+
+  arrow::Status Init() {
+    // If we are allowed to do delayed I/O, we can use a single request to upload the
+    // data. If not, we use a multi-part upload and initiate it here to
+    // sanitize that writing to the bucket is possible.
```

Comment:
nit: remove the comments since AFAIK the support of delayed IO is removed.

Reply:
fixed